staging_vespalib/src/vespa/vespalib/metrics/simple_metrics_manager.cpp

```diff
@@ -78,56 +77,53 @@ SimpleMetricsManager::mergeBuckets()
 }
 
 Snapshot
-SimpleMetricsManager::snapshot()
+SimpleMetricsManager::snapshotFrom(const Bucket &bucket)
 {
-    Bucket merged = mergeBuckets();
     std::vector<PointSnapshot> points;
 
-    std::chrono::microseconds s = since_epoch(merged.startTime);
-    std::chrono::microseconds e = since_epoch(merged.endTime);
+    std::chrono::microseconds s = since_epoch(bucket.startTime);
+    std::chrono::microseconds e = since_epoch(bucket.endTime);
     const double micro = 0.000001;
     Snapshot snap(s.count() * micro, e.count() * micro);
     {
-        Guard guard(_pointMaps.lock);
-        for (auto entry : _pointMaps.vec) {
-             const PointMapBacking &map = entry->first.backing();
+        for (size_t i = 0; i < _pointMaps.size(); ++i) {
+             const PointMap::BackingMap &map = _pointMaps.lookup(i).backingMap();
              PointSnapshot point;
-             for (const PointMapBacking::value_type &kv : map) {
+             for (const PointMap::BackingMap::value_type &kv : map) {
                  point.dimensions.emplace_back(nameFor(kv.first), valueFor(kv.second));
              }
              snap.add(point);
         }
     }
-    for (const CounterAggregator& entry : merged.counters) {
-        size_t ni = entry.idx.name_idx;
-        size_t pi = entry.idx.point_idx;
+    for (const CounterAggregator& entry : bucket.counters) {
+        size_t ni = entry.idx.name().id();
+        size_t pi = entry.idx.point().id();
         const vespalib::string &name = _metricNames.lookup(ni);
         CounterSnapshot val(name, snap.points()[pi], entry);
         snap.add(val);
     }
-    for (const GaugeAggregator& entry : merged.gauges) {
-        size_t ni = entry.idx.name_idx;
-        size_t pi = entry.idx.point_idx;
+    for (const GaugeAggregator& entry : bucket.gauges) {
+        size_t ni = entry.idx.name().id();
+        size_t pi = entry.idx.point().id();
         const vespalib::string &name = _metricNames.lookup(ni);
         GaugeSnapshot val(name, snap.points()[pi], entry);
         snap.add(val);
     }
     return snap;
 }
 
-void
-SimpleMetricsManager::doCollectLoop(SimpleMetricsManager *me)
+Snapshot
+SimpleMetricsManager::snapshot()
 {
-    const std::chrono::milliseconds jiffy{20};
-    const std::chrono::seconds oneSec{1};
-    while (!me->_stopFlag) {
-        std::this_thread::sleep_for(jiffy);
-        InternalTimeStamp now = now_stamp();
-        InternalTimeStamp::duration elapsed = now - me->_curTime;
-        if (elapsed >= oneSec) {
-            me->collectCurrentBucket();
-        }
-    }
+    Bucket merged = mergeBuckets();
+    return snapshotFrom(merged);
+}
+
+Snapshot
+SimpleMetricsManager::totalSnapshot()
+{
+    Guard guard(_bucketsLock);
```

Comment:
consider making a copy of the total bucket while holding the lock, then making the snapshot outside the critical region.